src/GraphQL/Subscription/ObservableExtensions.cs

```diff
@@ -0,0 +1,273 @@
+                    return;
+                Queue(QueueType.Data, _transformNext(value, _token), default);
+            }
+
+            public void OnError(Exception error)
+            {
+                if (_token.IsCancellationRequested)
+                    return;
+                Queue(QueueType.Error, default, _transformError(error, _token));
+            }
+
+            public void OnCompleted() => Queue(QueueType.Completion, default, default);
+
+            /// <summary>
+            /// Queues the specified event and if necessary starts watching for an event to complete.
+            /// </summary>
+            private void Queue(QueueType queueType, ValueTask<TOut> task, ValueTask<Exception> error)
+            {
+                if (_token.IsCancellationRequested)
+                    return; // note: the task to be queued is never awaited in this case
+                var queueData = new QueueEvent { Type = queueType, Data = task, Error = error };
+                bool attach = false;
+                lock (_queue)
+                {
+                    _queue.Enqueue(queueData);
+                    attach = _queue.Count == 1;
+                }
+
+                // start watching for an event to complete, if this is the first in the queue
+                if (attach)
+                {
+                    // start sending data events (will await on the task queued if needed)
+                    _ = ProcessAllEventsInQueueAsync();
+                }
+            }
+
+            /// <summary>
+            /// Processes data from the queue in order (or raises errors or completed notifications);
+            /// executes until the queue is empty.
+            /// </summary>
+            private async Task ProcessAllEventsInQueueAsync()
+            {
+                // grab the event at the start of the queue, but don't remove it from the queue
+                QueueEvent queueEvent;
+                bool moreEvents;
+                lock (_queue)
+                {
+                    // should always successfully peek from the queue here
+                    moreEvents = _queue.TryPeek(out queueEvent);
+                }
+                while (moreEvents)
+                {
+                    // process the event
+                    try
+                    {
+                        if (queueEvent.Type == QueueType.Data)
+                        {
+                            await ProcessDataAsync(queueEvent.Data).ConfigureAwait(false);
+                        }
+                        else if (queueEvent.Type == QueueType.Error)
+                        {
+                            await ProcessErrorAsync(queueEvent.Error).ConfigureAwait(false);
+                        }
+                        else if (queueEvent.Type == QueueType.Completion)
+                        {
+                            ProcessCompletion();
+                        }
+                    }
+                    catch { }
```

Comment:
Are you 100% sure?

Reply:
Yes.  If you see `ProcessDataAsync`, exceptions are caught and passed to `OnError`.  But if `OnError` throws an exception, this exception will bubble into `ProcessAllEventsInQueueAsync` and terminate execution of the method.  So then there will be the item in the queue that is never removed, and so future calls to `Queue` will never start another `ProcessAllEventsInQueueAsync`, and queued items will never be dequeued or dispatched.

Reply:
Good reason to mention in comment for empty `catch` block.

Reply:
May be add a static event to provide diagnostics for such cases? Like `TaskScheduler.UnobservedTaskException`.

Reply:
```suggestion
                    catch
                    {
                        // if an error is unhandled, execution of this method will terminate, and
                        // no further events in the queue will be processed, and Queue will not
                        // start another ProcessAllEventsInQueueAsync because the queue is not empty
                    }
```

Reply:
> May be add a static event to provide diagnostics for such cases? Like `TaskScheduler.UnobservedTaskException`.

Not a bad idea, but I am not inclined to add such a method.  If we did so, we would then need to wrap that inside a try catch block.

Reply:
Merged comment above.